services/src/main/java/io/grpc/services/BinaryLog.java

```diff
@@ -45,45 +48,120 @@
+        .newBuilder()
+        .setType(Type.SEND_INITIAL_METADATA)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setPeer(socketToProto(peerSocket))
+        .setMetadata(metadataToProto(metadata, maxHeaderBytes))
+        .build();
+    sink.write(entry);
+  }
+
+  /**
+   * Logs the trailing metadata. This method logs the appropriate number of bytes
+   * as determined by the binary logging configuration.
+   */
+  public void logTrailingMetadata(Metadata metadata, boolean isServer, byte[] callId) {
+    GrpcLogEntry entry = GrpcLogEntry
+        .newBuilder()
+        .setType(Type.SEND_TRAILING_METADATA)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setMetadata(metadataToProto(metadata, maxHeaderBytes))
+        .build();
+    sink.write(entry);
+  }
+
+  /**
+   * Logs the outbound message. This method logs the appropriate number of bytes from
+   * {@code message}, and returns an {@link InputStream} that contains the original message.
+   * The number of bytes logged is determined by the binary logging configuration.
+   */
+  public void logOutboundMessage(
+      byte[] message, boolean compressed, boolean isServer, byte[] callId) {
+    GrpcLogEntry entry = GrpcLogEntry
+        .newBuilder()
+        .setType(Type.SEND_MESSAGE)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setMessage(messageToProto(message, compressed, maxMessageBytes))
+        .build();
+    sink.write(entry);
+  }
+
+  /**
+   * Logs the inbound message. This method logs the appropriate number of bytes from
+   * {@code message}, and returns an {@link InputStream} that contains the original message.
+   * The number of bytes logged is determined by the binary logging configuration.
+   */
+  public void logInboundMessage(
+      byte[] message, boolean compressed, boolean isServer, byte[] callId) {
+    GrpcLogEntry entry = GrpcLogEntry
+        .newBuilder()
+        .setType(Type.RECV_MESSAGE)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setMessage(messageToProto(message, compressed, maxMessageBytes))
+        .build();
+    sink.write(entry);
+  }
+
   @Override
   public boolean equals(Object o) {
     if (!(o instanceof BinaryLog)) {
       return false;
     }
     BinaryLog that = (BinaryLog) o;
     return this.maxHeaderBytes == that.maxHeaderBytes
-        && this.maxMessageBytes == that.maxMessageBytes;
+        && this.maxMessageBytes == that.maxMessageBytes
+        && this.sink.equals(that.sink);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hashCode(maxHeaderBytes, maxMessageBytes);
+    return Objects.hashCode(maxHeaderBytes, maxMessageBytes, sink);
```

Comment:
Since sink is an abstract class, and not immutable, the hash code is not well defined.  maybe just leave it out?

Reply:
done